### ISSUE TITLE
Not Finding Classpath Elements on Java 11 

### ISSUE THREAD

**Issue by jamesward** [classgraph/classgraph#382]
I'm using ClassGraph 4.8.53 and having a problem with Java 11 not finding classpath elements but it's working on Java 8.

Here is what I'm seeing on Java 11:
```
2019-11-17T08:01:56.968-0700	ClassGraph	Finding classpath
2019-11-17T08:01:56.968-0700	ClassGraph	-- Finding classpath and modules
2019-11-17T08:01:56.969-0700	ClassGraph	---- Found ClassLoaders:
2019-11-17T08:01:56.969-0700	ClassGraph	------ jdk.internal.loader.ClassLoaders$AppClassLoader
2019-11-17T08:01:56.969-0700	ClassGraph	---- Overriding classpath with: jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd
2019-11-17T08:01:56.969-0700	ClassGraph	------ Found classpath element: jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd -> /home/jw/projects/webjars/webjars-play/jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd
2019-11-17T08:01:56.969-0700	ClassGraph	------ WARNING: when the classpath is overridden, there is no guarantee that the classes found by classpath scanning will be the same as the classes loaded by the context classloader
2019-11-17T08:01:56.969-0700	ClassGraph	-- Module scanning is disabled, because classloaders or classpath was overridden
2019-11-17T08:01:56.970-0700	ClassGraph	Opening classpath elements (took 0.026549 sec)
2019-11-17T08:01:56.970-0700	ClassGraph	-- Skipping invalid classpath element /home/jw/projects/webjars/webjars-play/jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd : java.io.FileNotFoundException
2019-11-17T08:01:56.996-0700	ClassGraph	Finding nested classpath elements
2019-11-17T08:01:56.996-0700	ClassGraph	Final classpath element order:
2019-11-17T08:01:56.997-0700	ClassGraph	Scanning classpath elements (took 0.000006 sec)
2019-11-17T08:01:56.997-0700	ClassGraph	Classfile scanning is disabled
```

This seems to only be a problem when I call `overrideClassLoaders(classLoader)`.  If I call `getResources` on the classloader I can see everything I'm looking for.

Sorry I don't understand much about how the classpath stuff might have changed in Java 11.

**Comment by lukehutch** [comment]
Hi @jamesward ,

Thanks for the report. I see one immediate issue based on the logs: you are calling `overrideClasspath(classLoader)`, not `overrideClassLoaders(classLoader)`. The reason I can tell that is that `overrideClasspath` calls `toString` on whatever it is passed (except arrays and lists -- those have `toString` called on all their elements). That explains this line:

```
Overriding classpath with: jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd
```

and subsequently this breaks, assuming that this is a directory or jar name, relative to the current directory:

```
/home/jw/projects/webjars/webjars-play/jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd
```

If you call `overrideClassLoaders` with a custom `ClassLoader`, then yes, only that `ClassLoader`'s paths will be scanned (which switches off module scanning). The reason for this is that in Java 9+, the standard `AppClassLoader` does not extend `URLClassLoader` anymore, and it has no API whatsoever for fetching its internal paths. Strong encapsulation means that even reflection cannot retrieve its internal paths. So the only way to scan the contents of `AppClassLoader` is to use the JPMS API: 

https://stackoverflow.com/questions/41932635/scanning-classpath-modulepath-in-runtime-in-java-9/45612376#45612376

Therefore if you override the context `AppClassLoader` with another custom non-module `ClassLoader`, you're actually overriding module scanning.

Hope that makes sense?

Do you need a way to scan both a custom `ClassLoader` *and* the current system- and non-system-modules? If so, I could add a call that re-enables module scanning when you call `overrideClassLoader`. Or you could run two separate scans, one for the module system, and one for the custom `ClassLoader`.

**Comment by jamesward** [comment]
Thanks Luke! You are way too kind to those of us that make silly mistakes.  I switched it back to use `overrideClassLoaders` (which I'd had before I started mucking with this.  And now I'm seeing something more reasonable:
```
INFO: 2019-11-17T18:47:17.411-0700	ClassGraph	ClassGraph version 4.8.53
2019-11-17T18:47:17.411-0700	ClassGraph	Operating system: Linux 5.3.0-23-generic amd64
2019-11-17T18:47:17.411-0700	ClassGraph	Java version: 11.0.5 / 11.0.5+10 (AdoptOpenJDK)
2019-11-17T18:47:17.411-0700	ClassGraph	Java home: /home/jw/bin/jdk-11.0.5+10
2019-11-17T18:47:17.417-0700	ClassGraph	ScanSpec:
2019-11-17T18:47:17.417-0700	ClassGraph	-- packageWhiteBlackList: whitelist: ["META-INF.resources.webjars"]; whitelistPrefixes: ["", "/", "META-INF.", "META-INF.resources.", "META-INF.resources.webjars."]
2019-11-17T18:47:17.417-0700	ClassGraph	-- packagePrefixWhiteBlackList: whitelistPrefixes: ["META-INF.resources.webjars."]
2019-11-17T18:47:17.417-0700	ClassGraph	-- pathWhiteBlackList: whitelist: ["META-INF/resources/webjars/"]; whitelistPrefixes: ["", "/", "META-INF/", "META-INF/resources/", "META-INF/resources/webjars/"]
2019-11-17T18:47:17.417-0700	ClassGraph	-- pathPrefixWhiteBlackList: whitelistPrefixes: ["META-INF/resources/webjars/"]
2019-11-17T18:47:17.417-0700	ClassGraph	-- classWhiteBlackList: 
2019-11-17T18:47:17.417-0700	ClassGraph	-- classfilePathWhiteBlackList: 
2019-11-17T18:47:17.417-0700	ClassGraph	-- classPackageWhiteBlackList: 
2019-11-17T18:47:17.417-0700	ClassGraph	-- classPackagePathWhiteBlackList: 
2019-11-17T18:47:17.417-0700	ClassGraph	-- moduleWhiteBlackList: 
2019-11-17T18:47:17.417-0700	ClassGraph	-- jarWhiteBlackList: 
2019-11-17T18:47:17.418-0700	ClassGraph	-- classpathElementResourcePathWhiteBlackList: 
2019-11-17T18:47:17.418-0700	ClassGraph	-- libOrExtJarWhiteBlackList: 
2019-11-17T18:47:17.418-0700	ClassGraph	-- performScan: true
2019-11-17T18:47:17.418-0700	ClassGraph	-- scanJars: true
2019-11-17T18:47:17.418-0700	ClassGraph	-- scanNestedJars: true
2019-11-17T18:47:17.418-0700	ClassGraph	-- scanDirs: true
2019-11-17T18:47:17.418-0700	ClassGraph	-- scanModules: true
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableClassInfo: false
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableFieldInfo: false
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableMethodInfo: false
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableAnnotationInfo: false
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableStaticFinalFieldConstantInitializerValues: false
2019-11-17T18:47:17.418-0700	ClassGraph	-- enableInterClassDependencies: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- enableExternalClasses: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- enableSystemJarsAndModules: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- ignoreClassVisibility: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- ignoreFieldVisibility: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- ignoreMethodVisibility: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- disableRuntimeInvisibleAnnotations: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- extendScanningUpwardsToExternalClasses: true
2019-11-17T18:47:17.419-0700	ClassGraph	-- enableRemoteJarScanning: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- addedClassLoaders: null
2019-11-17T18:47:17.419-0700	ClassGraph	-- overrideClassLoaders: [jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd]
2019-11-17T18:47:17.419-0700	ClassGraph	-- addedModuleLayers: null
2019-11-17T18:47:17.419-0700	ClassGraph	-- overrideModuleLayers: null
2019-11-17T18:47:17.419-0700	ClassGraph	-- overrideClasspath: null
2019-11-17T18:47:17.419-0700	ClassGraph	-- classpathElementFilters: null
2019-11-17T18:47:17.419-0700	ClassGraph	-- initializeLoadedClasses: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- removeTemporaryFilesAfterScan: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- ignoreParentClassLoaders: true
2019-11-17T18:47:17.419-0700	ClassGraph	-- ignoreParentModuleLayers: false
2019-11-17T18:47:17.419-0700	ClassGraph	-- modulePathInfo: 
2019-11-17T18:47:17.420-0700	ClassGraph	Number of worker threads: 14
2019-11-17T18:47:17.423-0700	ClassGraph	Finding classpath
2019-11-17T18:47:17.423-0700	ClassGraph	-- Finding classpath and modules
2019-11-17T18:47:17.424-0700	ClassGraph	---- Override ClassLoaders:
2019-11-17T18:47:17.424-0700	ClassGraph	------ jdk.internal.loader.ClassLoaders$AppClassLoader
2019-11-17T18:47:17.424-0700	ClassGraph	---- System jars:
2019-11-17T18:47:17.425-0700	ClassGraph	------ Scanning disabled for lib or ext jar: /home/jw/bin/jdk-11.0.5+10/lib/jrt-fs.jar
2019-11-17T18:47:17.425-0700	ClassGraph	---- ClassLoaderHandlers:
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.AntClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.EquinoxContextFinderClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.FelixClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JBossClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WeblogicClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereLibertyClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.WebsphereTraditionalClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.OSGiDefaultClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.SpringBootRestartClassLoaderHandler
2019-11-17T18:47:17.431-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.TomcatWebappClassLoaderBaseHandler
2019-11-17T18:47:17.432-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.PlexusClassWorldsClassRealmClassLoaderHandler
2019-11-17T18:47:17.432-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.ParentLastDelegationOrderTestClassLoaderHandler
2019-11-17T18:47:17.432-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2019-11-17T18:47:17.432-0700	ClassGraph	------ nonapi.io.github.classgraph.classloaderhandler.URLClassLoaderHandler
2019-11-17T18:47:17.432-0700	ClassGraph	---- Finding unique classloaders in delegation order
2019-11-17T18:47:17.433-0700	ClassGraph	---- Obtaining URLs from classloaders in delegation order
2019-11-17T18:47:17.433-0700	ClassGraph	------ Ignoring parent classloader jdk.internal.loader.ClassLoaders$PlatformClassLoader@448f790, normally handled by nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2019-11-17T18:47:17.433-0700	ClassGraph	------ Classloader jdk.internal.loader.ClassLoaders$AppClassLoader@77556fd is handled by nonapi.io.github.classgraph.classloaderhandler.JPMSClassLoaderHandler
2019-11-17T18:47:17.433-0700	ClassGraph	-- Module scanning is disabled, because classloaders or classpath was overridden
2019-11-17T18:47:17.434-0700	ClassGraph	Opening classpath elements (took 0.000823 sec)
2019-11-17T18:47:17.435-0700	ClassGraph	Finding nested classpath elements
2019-11-17T18:47:17.435-0700	ClassGraph	Final classpath element order:
2019-11-17T18:47:17.435-0700	ClassGraph	Scanning classpath elements (took 0.000151 sec)
```

And I think that based on your answer I'll have to try both module scanning & classloader scanning since I need to support Java 8 & Java 11.  I'll let you know what happens as I give it a try.

**Comment by jamesward** [comment]
Here is what I'm trying:
https://github.com/webjars/webjars-locator-core/commit/a58ac5611a8f9d25cc67d2a033d0fb476f2e09c7

Let me know if that is terrible.  Thanks!

**Comment by lukehutch** [comment]
I don't think you need to scan twice. If this method is called:

https://github.com/webjars/webjars-locator-core/blob/a58ac5611a8f9d25cc67d2a033d0fb476f2e09c7/src/main/java/org/webjars/WebJarAssetLocator.java#L162

then the scan will work fine for either JDK 7/8 or JDK 9+. (It will automatically scan modules by default if you're running on JDK 9+.)

If instead you pass in a custom classloader:

https://github.com/webjars/webjars-locator-core/blob/a58ac5611a8f9d25cc67d2a033d0fb476f2e09c7/src/main/java/org/webjars/WebJarAssetLocator.java#L166

then as long as you're not passing in the current context classloader or the bootstrap loader in JDK 9+ (which cannot actually be scanned directly, since the classpath cannot be obtained from those classloaders), again, it should work for all JDK versions. (For example if you create a new `URLClassLoader` and pass a ref to that in, it will scan fine on any JDK.) There's no reason to pass in the context classloader, because simply not overriding the classloader will already scan the context classloader (or module system), so you would only ever call this method if you had created a custom classloader.

If you're worried about efficiency, and only want to scan jars that contain a file of interest, e.g. `pom.properties`:

https://github.com/webjars/webjars-locator-core/blob/a58ac5611a8f9d25cc67d2a033d0fb476f2e09c7/src/main/java/org/webjars/WebJarAssetLocator.java#L94

then you could try calling `whitelistClasspathElementsContainingResourcePath(paths)` before calling `scan()`. This will open all classpath elements, but do no more than read the directory or zipfile entries for the classpath elements to see if they contain the required paths, and abort scanning if not. It will be a bit faster if you have a large classpath.

**Comment by jamesward** [comment]
Thanks Luke for the details.  In my case there are a bunch of ways that this gets used.  Sometimes there isn't a custom classloader.  Sometimes there is, and in that case it could be anything.  Right now my troubles are because Play Framework when run by sbt seems to be passing in an `AppClassLoader` and ClassGraph can't find the resources in it on Java 11.

**Comment by lukehutch** [comment]
Ah, ok... In that case, if an `AppClassLoader` is passed in, the solution in that case is just to ignore it, and call ClassGraph without overriding the classloader. If you want to be doubly sure, check if the passed in classloader is equal to the context classloader. It should be, since (as far as I know) there's no way to create your own `AppClassLoader` in JDK9+, and there should only ever be one instance of it. In other words, if it's passed in, it's simply obtained by Play by getting the context classloader. Therefore, ignore it and don't override it.

**Comment by jamesward** [comment]
Thanks!  This seems to be working:
https://github.com/webjars/webjars-play/blob/master/src/main/scala/org/webjars/play/WebJarsUtil.scala#L90-L99

**Comment by lukehutch** [comment]
Looks right to me. If there are any corner cases that are still broken, feel free to comment here.

Actually though, this raises a good point... Maybe that exact logic should be moved to ClassGraph, so that nobody else runs into this. I'll work on doing that.

**Comment by jamesward** [comment]
Really appreciate all your prompt support!

**Comment by lukehutch** [comment]
You're welcome, that's the goal! I try to maintain this project how I wish all open source software was maintained. (This is the software version of The Golden Rule :-) )

The change I just checked in catches one additional case -- it also enables system module scanning if you override the classloaders with the parent of an `AppClassLoader`, which is a `PlatformClassLoader` (which only loads classes for system modules).

It's a simple enough change that I'll just cross my fingers and hope it works :-D  Released in 4.8.54. You should be able to revert your latest change and just depend upon this version to fix the problem.

Thanks for your report!

**Comment by lukehutch** [comment]
The fix for this was incomplete, see #411. The full fix is in git master, and will be released in 4.8.67 (might take a few days to hit Maven Central, because Sonatype is currently having issues).

**Comment by lukehutch** [comment]
I found a workaround for the Sonatype issue, so the fix is now published in classgraph-4.8.67.

**Comment by jamesward** [comment]
Awesome. Thank you!
